File: extension/host/keygrouper.py
class Part:
    # text could be pasted text or initial value
    def __init__(self, start, end, _id, text):
        self.start = start
        self.end = end
        self.raw_end = len(text) if end == "E" else end
        self.id = _id
        self.text = text

    def indicesWithinBoundaries(self):
        outOfBounds = self.start < 0 or self.raw_end < 0 or self.raw_end <= self.start
        return not outOfBounds

    def getVal(self):
        if not self.indicesWithinBoundaries():
            return ""
        return self.text[self.start:self.raw_end]
    def __str__(self):
        return str([self.start,self.end,self.raw_end, self.text, self.id])
    def __eq__(self, other):
        """Overrides the default implementation"""
        if isinstance(other, Part):
            return self.start == other.start and \
            self.end == other.end and \
            self.raw_end == other.raw_end and \
            self.id == other.id and \
            self.text == other.text
        return False
    def __ne__(self, other):
        """Overrides the default implementation (unnecessary in Python 3)"""
        return not self.__eq__(other)
    def __hash__(self):
        """Overrides the default implementation"""
        return hash(tuple(sorted(self.__dict__.items())))
    def __repr__(self):
        return self.__str__()

class KeyGrouper:
    def __init__(self, initialValue=""):
        self.initiaValueWasSet = False
        self.parts = []
        if len(initialValue) > 0:
            self.initiaValueWasSet = True
            self.parts = [Part(0, "E", "INITIAL_VALUE", initialValue)]
    
    def getVal(self):
        val = ""
        for part in self.parts:
            val += self.getValForPart(part)
        return val
    
    def getValForPart(self, part):
        val = ""
        if type(part) == str:
            val = part
        elif type(part) == Part:
            return part.getVal()
        return val
    
    # TODO concatenate consecutive parts if for example (0,1) (1,2) on same paste
    # TODO later even smarter, if user removes char from paste then types it again
    def getParts(self):
        if len(self.parts) == 0:
            return []
        finalParts = [self.parts[0]]
        for i in range(1, len(self.parts)):
            part = self.parts[i]
            if type(part) == str and type(finalParts[-1]) == str:
                finalParts[-1] += part
            elif (type(part) == str and type(finalParts[-1]) == Part) or type(part) == Part:
                finalParts.append(part)
        return finalParts

    def printParts(self):
        s = ""
        for part in self.parts:
            s+=str(part)
        return s

    def getLenPart(self, i):
        return len(self.getValForPart(self.parts[i]))
    
    # If initialValue is here i can check how much of it was removed/split..etc
    # if its not there we cant tell if it wasnt there from the first place or if it was there
    # then removed
    def wasInitialValueRemoved(self):
        return not any([type(p) == Part and p.id == "INITIAL_VALUE" for p in self.parts]) and \
            self.initiaValueWasSet
    def getPartToUpdateOrIndexForOffset(self, offset):
        currentOffset = 0
        for i in range(len(self.parts)):
            currentOffset += self.getLenPart(i)
            part = self.parts[i]
            if currentOffset > offset and type(self.parts[i]) == str:
                return None, i
            elif currentOffset > offset:
                return self.getLenPart(i) - (currentOffset- offset), i
        return None, currentOffset
    def appendTextAtOffset(self, char, offset):
        for c in reversed(char):
            self._appendTextAtOffset(c, offset)
    def _appendTextAtOffset(self, char, offset):
        indexInsidePart, index = self.getPartToUpdateOrIndexForOffset(offset)
        # if indexInsidePart is 0 then its the beginning of the part, we can just append it before it
        if indexInsidePart is not None and indexInsidePart != 0:
            p = self.parts[index]
            p1 = Part(p.start, p.start + indexInsidePart, p.id, p.text)
            p2 = Part(p.start + indexInsidePart, p.end, p.id, p.text)
            self.parts[index] = p1
            self.parts.insert(index + 1, char)
            self.parts.insert(index + 2, p2)
        elif index is not None:
            self.parts.insert(index, char)
    # TODO: allow more than one paste
    def appendPasteAtOffset(self, pastedText, offset, pasteId):
        indexInsidePart, index = self.getPartToUpdateOrIndexForOffset(offset)
        if index is not None:
            part = Part(0, "E", pasteId, pastedText)
            self.parts.insert(index, part)
    
    def deleteTextAtOffset(self, offset):
        if len(self.parts) == 0:
            return
        indexInsidePart, index = self.getPartToUpdateOrIndexForOffset(offset)
        if indexInsidePart is not None:
            p = self.parts[index]
            p1 = Part(p.start, p.start + indexInsidePart, p.id, p.text)
            p2 = Part(p.start + indexInsidePart + 1, p.end, p.id, p.text) 
            if len(self.getValForPart(p1)) != 0:
                self.parts[index] = p1
            if len(self.getValForPart(p2)) != 0:
                if len(self.getValForPart(p1)) != 0 :
                    self.parts.insert(index+1, p2)
                else:
                    self.parts[index] = p2
            if len(self.getValForPart(p1)) == 0 and len(self.getValForPart(p2)) == 0:
                del self.parts[index]
        elif index is not None:
            del self.parts[index]
    
    def deleteTextAtOffsetRange(self, startOffset, endOffset):
        for i in range(startOffset, endOffset):
            self.deleteTextAtOffset(startOffset)
    
    def __eq__(self, other):
        """Overrides the default implementation"""
        if isinstance(other, KeyGrouper):
            return self.initiaValueWasSet == other.initiaValueWasSet and \
                self.parts == other.parts
        return False
    
    def __ne__(self, other):
        """Overrides the default implementation (unnecessary in Python 3)"""
        return not self.__eq__(other)

    def __str__(self):
        return str(self.printParts())+"||"+self.getVal()

    def __repr__(self):
        return self.__str__()

File: extension/hostjs/keygrouper.js
class Part {
    // text could be pasted text or initial value
    constructor(start, end, _id, text) {
        this.start = start;
        this.end = end;
        this.raw_end = end == "E" ? text.length : end;
        this.id = _id;
        this.text = text;
    }

    indicesWithinBoundaries() {
        let outOfBounds = this.start < 0 || this.raw_end < 0 || this.raw_end <= this.start;
        return !outOfBounds;
    }

    getVal() {
        if (!this.indicesWithinBoundaries()) {
            return "";
        }
        return this.text.slice(this.start, this.raw_end);
    }
    jsonify() {
        return {
            "start": this.start,
            "end": this.end,
            "id": this.id,
        };
    }
    toString() {
        return String([this.start, this.end, this.raw_end, this.text, this.id]);
    }
    equals(other) {
        if (other instanceof Part) {
            return this.start == other.start &&
                this.end == other.end &&
                this.raw_end == other.raw_end &&
                this.id == other.id &&
                this.text == other.text;
        }
        return false;
    }
}

class KeyGrouper {
    constructor(initialValue = "") {
        this.initialValueWasSet = false;
        this.parts = [];
        if (initialValue.length > 0) {
            this.initialValueWasSet = true;
            this.parts = [new Part(0, "E", "INITIAL_VALUE", initialValue)];
        }
    }

    getVal() {
        let val = "";
        for (let part of this.parts) {
            val += this.getValForPart(part);
        }
        return val;
    }

    getValForPart(part) {
        let val = "";
        if (typeof part == "string") {
            val = part;
        } else if (part instanceof Part) {
            return part.getVal();
        }
        return val;
    }
    jsonify() {
        let parts = this.getParts();
        let jsonExport = [];
        for (let p of parts) {
            if (typeof p == "string") {
                jsonExport.push(p);
            } else {
                jsonExport.push(p.jsonify());
            }
        }
        return jsonExport;
    }

    // TODO concatenate consecutive parts if for example (0,1) (1,2) on same paste
    // TODO later even smarter, if user removes char from paste then types it again
    getParts() {
        if (this.parts.length == 0) {
            return [];
        }
        // Consecutive strings concatenation
        let finalParts = [this.parts[0]];
        for (let i = 1; i < this.parts.length; i++) {
            let part = this.parts[i];
            if (typeof part == 'string' && typeof finalParts[finalParts.length - 1] == 'string') {
                finalParts[finalParts.length - 1] += part;
            } else if ((typeof part == 'string' && finalParts[finalParts.length - 1] instanceof Part) || part instanceof Part) {
                finalParts.push(part);
            }
        }
        return finalParts;
    }

    printParts() {
        let s = "";
        for (let part of this.parts) {
            s += String(part);
        }
        return s;
    }

    getLenPart(i) {
        return this.getValForPart(this.parts[i]).length;
    }

    // If initialValue is here i can check how much of it was removed/split..etc
    // if its not there we cant tell if it wasnt there from the first place or if it was there
    // then removed
    wasInitialValueRemoved() {
        let isThereAnInitialValue = !!this.parts.filter(p => p instanceof Part && p.id == "INITIAL_VALUE").length;
        return !isThereAnInitialValue && this.initialValueWasSet;
    }
    getPartToUpdateOrIndexForOffset(offset) {
        let currentOffset = 0;
        for (let i in this.parts) {
            currentOffset += this.getLenPart(i);
            if (currentOffset > offset && typeof this.parts[i] == "string") {
                return { indexInsidePart: null, index: parseInt(i) };
            } else if (currentOffset > offset) {
                return { indexInsidePart: this.getLenPart(i) - (currentOffset - offset), index: parseInt(i) };
            }
        }
        return { indexInsidePart: null, index: currentOffset };
    }
    appendTextAtOffset(text, offset) {
        for (let c of text.split('').reverse().join('')) {
            this._appendTextAtOffset(c, offset);
        }
    }
    _appendTextAtOffset(char, offset) {
        let { indexInsidePart, index } = this.getPartToUpdateOrIndexForOffset(offset);
        // if indexInsidePart is 0 then its the beginning of the part, we can just append it before it
        if (indexInsidePart != null && indexInsidePart != 0) {
            let p = this.parts[index];
            let p1 = new Part(p.start, p.start + indexInsidePart, p.id, p.text);
            let p2 = new Part(p.start + indexInsidePart, p.end, p.id, p.text);
            this.parts[index] = p1;
            this.parts.splice(index + 1, 0, char);
            this.parts.splice(index + 2, 0, p2);
        } else {
            this.parts.splice(index, 0, char);
        }
    }
    // TODO: allow more than one paste
    appendPasteAtOffset(pastedText, offset) {
        let { _, index } = this.getPartToUpdateOrIndexForOffset(offset);
        let part = new Part(0, "E", "PASTE", pastedText);
        this.parts.splice(index, 0, part);
    }

    deleteTextAtOffset(offset) {
        if (this.parts.length == 0) {
            return;
        }
        let { indexInsidePart, index } = this.getPartToUpdateOrIndexForOffset(offset);
        if (indexInsidePart != null) {
            let p = this.parts[index];
            let p1 = new Part(p.start, p.start + indexInsidePart, p.id, p.text);
            let p2 = new Part(p.start + indexInsidePart + 1, p.end, p.id, p.text);
            if (this.getValForPart(p1).length != 0) {
                this.parts[index] = p1;
            }
            if (this.getValForPart(p2).length != 0) {
                if (this.getValForPart(p1).length != 0) {
                    this.parts.splice(index + 1, 0, p2);
                } else {
                    this.parts[index] = p2;
                }
            }
            if (this.getValForPart(p1).length == 0 && this.getValForPart(p2).length == 0) {
                this.parts.splice(index, 1); // deletes index from array;
            }
        } else {
            this.parts.splice(index, 1); // deletes index from array;
        }
    }

    deleteTextAtOffsetRange(startOffset, endOffset) {
        for (let i = startOffset; i < endOffset; i++) {
            this.deleteTextAtOffset(startOffset);
        }
    }

    equals(other, checkInitialValueWasSet) {
        if (other instanceof KeyGrouper) {
            if (checkInitialValueWasSet && this.initialValueWasSet != other.initialValueWasSet) {
                return false;
            }
            return this.jsonify() == other.jsonify();
        }
        return false;
    }

    toString() {
        return this.printParts() + "||" + this.getVal();
    }
}

module.exports = {
    KeyGrouper: KeyGrouper,
    Part: Part,
};